New Kafka input operator using 0.9.0 Kafka consumer API

siyuanh opened the pull request:
Some highlights of the operator
1. Out-of-box One-to-one and one-to-many partition scheme support plus customizable partition schema
2. Fault-tolerant when the input operator goes down, it redeploys on other node
3. At-least-once semantics for operator failure (no matter which operator fails)
4. At-least-once semantics for cold restart (no data loss even if you restart the application)
5. Multi-cluster support, one operator can consume data from more than one kafka clusters
6. Multi-topic support, one operator can subscribe multiple topics
7. Throughput control support, you can throttle number of tuple for each streaming window

Things needs to be done,
More unit test for different partition scheme, application restart
More integration test
Metrics for operator
Bring back idempotency

tweise commented on kafka/pom.xml:
```diff
@@ -0,0 +1,278 @@
+  </properties>
+  <repositories>
+    <repository>
+      <id>datatorrent-3rd-party</id>
```
What is this needed for?

tweise commented on kafka/pom.xml:
```diff
@@ -0,0 +1,278 @@
+        </executions>
+      </plugin>
+     <!-- create resource directory for xml javadoc-->
+     <plugin>
```
We should find a way to not repeat this in every project.

tweise commented on kafka/pom.xml:
```diff
@@ -0,0 +1,278 @@
+      <artifactId>commons-lang3</artifactId>
+      <version>3.1</version>
+    </dependency>
+    <dependency>
```
Isn't this an engine dependency?

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+/**
+ * The abstract kafka input operator using kafka 0.9.0 new consumer API
+ * A scalable, fault-tolerant, at-least-once kafka input operator
+ * Keu features includes:
```
Typo

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+ * The abstract kafka input operator using kafka 0.9.0 new consumer API
+ * A scalable, fault-tolerant, at-least-once kafka input operator
+ * Keu features includes:
+ * 1. Out-of-box One-to-one and one-to-many partition schema support plus customizable partition schem
```
Typo

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+ */
+public abstract class AbstractKafkaInputOperator<K, V> implements InputOperator, Operator.ActivationListener<Context.OperatorContext>, Operator.CheckpointListener, Partitioner<AbstractKafkaInputOperator>, StatsListener
+{
+
```
extra blank lines

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaPartitioner.java:
```diff
@@ -0,0 +1,277 @@
+    output.close();
+    Input lInput = new Input(bos.toByteArray());
+    @SuppressWarnings("unchecked")
+    Partitioner.Partition<AbstractKafkaInputOperator> p = new DefaultPartition<>(kryo.readObject(lInput, prototypeOperator.getClass()));
```
Getting a compile error inside Eclipse, goes away with:

```
    Partitioner.Partition<AbstractKafkaInputOperator> p = (DefaultPartition<AbstractKafkaInputOperator>)new DefaultPartition<>(kryo.readObject(lInput, prototypeOperator.getClass()));

```

SuppressWarnings can be removed with this change also.

tweise replied:
There are actually raw type warnings also.

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaPartitioner.java:
```diff
@@ -0,0 +1,277 @@
+  public static class PartitionMeta
+  {
+
+    public PartitionMeta()
```
Why public?

siyuanh replied:
It is used as return type for assign method in AbstractKafkaPartitioner class

tweise replied:
I don't find any reference to the constructor.

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+    APPLICATION_OR_LATEST // consume from committed position from last run or latest if there is no committed offset(s)
+  }
+  /**
+   *  node separate by ',' and clusters separate by ';'
```
Add more description.

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+    if (repartitionInterval < 0 || repartitionCheckInterval < 0 ||
+      t - lastCheckTime < repartitionCheckInterval || t - lastRepartitionTime < repartitionInterval) {
+      // return false if it's within repartitionCheckInterval since last time it check the stats
+      logger.debug("Within Repartition Check Interval {} or Repartition Interval {}", repartitionCheckInterval,
```
No need to log on every call.

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+    return topics;
+  }
+
+  public void setConsumerWrapper(KafkaConsumerWrapper consumerWrapper)
```
raw type warning, occurs elsewhere also

tweise commented on kafka/pom.xml:
```diff
@@ -0,0 +1,278 @@
+  </parent>
+
+  <artifactId>malhar-kafka</artifactId>
+  <name>Apache Apex Malhar (incubating) Contrib Library</name>
```
Contrib Library -> Kafka Support

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+   * initial partition count
+   * only used with PartitionStrategy.ONE_TO_MANY
+   */
+  private int initialPartitionCount = 1;
```
Why is this not a partitioner property?

siyuanh replied:
I want this to be easily set on Operator level, actually I want to pull ConsumerWrapper properties to Operator as well. I think that is a easier way for developers who use this operator

tweise replied:
This does not work when it is the property of another property (with configuration, you cannot assume the order in which properties are set). If the partitioner won't be a property then it may work.

tweise commented on kafka/src/main/java/com/datatorrent/contrib/kafka090/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,410 @@
+  private int initialPartitionCount = 1;
+
+  // Minimal interval between 2 (re)partition actions
+  private long repartitionInterval = 30000L;
```
Should also go to the partitioner?

gauravgopi123 commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaPartitioner.java:
```diff
@@ -0,0 +1,276 @@
+  public Response processStats(BatchedOperatorStats batchedOperatorStats)
+  {
+    Response response = new Response();
+    response.repartitionRequired = true;
```
Do you always want to re-partition when processStats is called??

gauravgopi123 commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,455 @@
+    try {
+      logger.debug("Process stats");
+      initPartitioner();
+      return partitioner.processStats(batchedOperatorStats);
```
Correct me if I am wrong, but from this it looks like after repartitionInterval the repartitioning will always happen as partitioner.processStats evaluates to AbstractKafkaPartitioner.processStats and there repartitionRequired = true

siyuanh replied:
3 possibilities here
1. within interval, the operator shouldn't not try repartition
2. exceed interval, the partitioner will always try to repartition based on latest kafka partition info. If there is no kafka partition change, it will return same collection of partition instances and AM will not try to repartition the operator.
3. exceed interval and there is kafka partition change(ex. new kafka partition added at runtime) it will try to repartition the operator.

tweise commented on kafka/pom.xml:
```diff
@@ -0,0 +1,235 @@
+
+  <properties>
+    <!-- skip tests by default as they depend on manual setup -->
+    <skip.kafka.module.tests>false</skip.kafka.module.tests>
```
Why are tests skipped?

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaPartitioner.java:
```diff
@@ -0,0 +1,276 @@
+    kryo.writeObject(output, prototypeOperator);
+    output.close();
+    Input lInput = new Input(bos.toByteArray());
+    Partitioner.Partition<AbstractKafkaInputOperator> p = new DefaultPartition<>(kryo.readObject(lInput, prototypeOperator.getClass()));
```
Still getting the compile error in Eclipse and code is still littered with type warnings.

gauravgopi123 commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaPartitioner.java:
```diff
@@ -0,0 +1,276 @@
+        result.add(createPartition(partitionAssignment));
+        i++;
+      }
+
```
extra line break

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java:
```diff
@@ -0,0 +1,408 @@
+    @Override
+    public void onComplete(Map<TopicPartition, OffsetAndMetadata> map, Exception e)
+    {
+      if (logger.isDebugEnabled()) {
```
What's the purpose of this implementation?

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java:
```diff
@@ -0,0 +1,408 @@
+    this.cacheSize = cacheSize;
+  }
+
+  public void setOffsetCommitCallback(OffsetCommitCallback offsetCommitCallback)
```
Can't find a reference, what's the purpose of this API?

tweise commented on kafka/src/test/java/org/apache/apex/malhar/kafka/KafkaInputOperatorTest.java:
```diff
@@ -0,0 +1,221 @@
+
+  }
+
+  static final org.slf4j.Logger logger = LoggerFactory.getLogger(KafkaInputOperatorTest.class);
```
Why not private?

tweise commented on kafka/src/test/java/org/apache/apex/malhar/kafka/KafkaInputOperatorTest.java:
```diff
@@ -0,0 +1,221 @@
+    p.close();
+    lc.shutdown();
+    // kafka has a bug shutdown connector you have to make sure kafka client resource has been cleaned before clean the broker
+    Thread.sleep(5000);
```
Check for that condition instead?

siyuanh replied:
Not needed any more, copied from the old code

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,455 @@
+    return maxTuplesPerWindow;
+  }
+
+  public void setKeyDeserializer(Class<? extends Deserializer<K>> keyDeserializer)
```
Shouldn't the serializers be protected instead? The derived class has to provide them, but not the user?

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java:
```diff
@@ -0,0 +1,408 @@
+      if (ownerOperator.getKeyDeserializer() != null) {
+        prop.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ownerOperator.getKeyDeserializer().getCanonicalName());
+      }
+      if (ownerOperator.getValueDeserializer() != null) {
```
Wrap the serializer to be able to count the bytes received for bandwidth control or are there other options to accomplish that?

siyuanh replied:
We can have the consumer metrics from the consumer API. But if we want to control the bandwidth bytes/window ex., there is no explicit way to do that. One thing I can think of is we always use ByteArrayDeserializer and use the actual Deserializer to deserialize the message in operator code, but I don't know if it's a right idea.

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,512 @@
+    return topics;
+  }
+
+  public void setConsumerWrapper(KafkaConsumerWrapper consumerWrapper)
```
Why expose this as property?

siyuanh replied:
There are 2 extra properties you can set on the wrapper, but I think it's better to move them out to Operator so users can easily set the value and we can make this wrapper object transient

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java:
```diff
@@ -0,0 +1,378 @@
+    return initialOffset;
+  }
+
+  public int getCacheSize()
```
We should correct the property name.

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java:
```diff
@@ -0,0 +1,378 @@
+ *
+ * It also use the consumers to commit the application processed offsets along with the application name
+ *
+ * //@param The key object, needs to be used with http://kafka.apache.org/090/javadoc/org/apache/kafka/clients/consumer/ConsumerConfig.html#KEY_DESERIALIZER_CLASS_CONFIG
```
Remove this?

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,512 @@
+  private long repartitionCheckInterval = 5000L;
+
+  // class name of the key deserializer
+  protected String keyDeserializer;
```
Why String, why not objects? However, given that we convert in emitTuples, we probably don't need these at all.

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,512 @@
+      Pair<String, ConsumerRecord<byte[], byte[]>> tuple = consumerWrapper.pollMessage();
+      ConsumerRecord<byte[], byte[]> msg = tuple.getRight();
+      emitTuple(tuple.getLeft(), msg.topic(), msg.partition(), msg.offset(),
+          keyDeser.deserialize(msg.topic(), msg.key()), valueDeser.deserialize(msg.topic(), msg.value()));
```
Should leave conversion to emitTuple and derived class. This would follow the pattern used in other operators.

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,500 @@
+    this.initialOffset = InitialOffset.valueOf(initialOffset.toUpperCase());
+  }
+
+  public InitialOffset getInitialOffset()
```
Should be JavaBean compliant getter/setter.

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,500 @@
+    return initialOffset;
+  }
+
+  public String getApplicationName()
```
Should be protected scope.

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java:
```diff
@@ -0,0 +1,500 @@
+    this.consumerTimeout = consumerTimeout;
+  }
+
+  public int getHoldingBufferSize()
```
Maybe name this "capacity" instead?

tweise replied:
getQueueCapacity() ?

tweise commented on kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java:
```diff
@@ -0,0 +1,326 @@
+        prop.put(ConsumerConfig.GROUP_ID_CONFIG, ownerOperator.getApplicationName() + "_Consumer");
+      }
+
+      KafkaConsumer kc = new KafkaConsumer(prop);
```
rawtype warnings, elsewhere also

siyuanh committed:
APEXMALHAR-1904 #comment
  Add new KafkaInputOperator using kafka 0.9.x consumer API
  Features includes:
      Out-of-box One-to-one and one-to-many partition scheme support plus customizable partition schema
      Fault-tolerant when the input operator goes down, it redeploys on other node
      At-least-once semantics for operator failure (no matter which operator fails)
      At-least-once semantics for cold restart (no data loss even if you restart the application)
      Multi-cluster support, one operator can consume data from more than one kafka clusters
      Multi-topic support, one operator can subscribe multiple topics
      Throughput control support, you can throttle number of tuple for each streaming window
